### ISSUE TITLE
Error running forever.js on Windows - “'C:\Program' is not recognized as an internal or external command, operable program or batch file”

### ISSUE THREAD

**Issue by sambragg** [foreversd/forever#1056]
I am trying to run ```forever.js``` via the Windows command prompt and I get the following output:

```>npm i -g forever```

```/my-project>forever start index.js```

*Log output:*

```'C:\Program' is not recognized as an internal or external command, operable program or batch file``` 

I think it's something to do with the ```PATH``` that ```forever``` is using for the ```node``` binary, but I don't know how to fix it...

[Link](https://stackoverflow.com/questions/59710358/error-running-forever-js-on-windows-c-program-is-not-recognized-as-an-inte) to my question on Stack Overflow

**Comment by MrDj200** [comment]
I've got the same problem.
My best guess its the space in the path

**Comment by sunk818** [comment]
@sambragg If you type `dir /x` in the root of c:, you can see the [short name][1] of the directory.

So, try the following instead:

    C:\PROGRA~1\nodejs\node.exe

[![dir /x example][2]][2]


  [1]: https://web.archive.org/web/20190807225855/https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/dir
  [2]: https://i.stack.imgur.com/Z2WAz.png

**Comment by sambragg** [comment]
@sunk818 thanks for your reply on this issue and on SO. I understand the issue now thanks to your explanation but how do I tell `forever` to use this new path that you have suggested?

**Comment by sunk818** [comment]
@sambragg I've never run node npm on Windows so I don't know from personal experience. 

https://stackoverflow.com/questions/27864040/fixing-npm-path-in-windows-8-and-10

if you run `set` from the command prompt, one of the entries might be for nodejs. Then you can try editing the path with the short name. User environmental variables may require you to reboot your system (or at least log out). System environment variables should apply immediately to the next command prompt you open.

The other brute force way is run Process Monitor from sysinternals , filter on process name containing forever , and see how the software traverses through finding npm. I think getting stuff like this to run in Windows is painful. At least consider running on Ubuntu WSL (Windows for Subsystem Linux) in the future and see if you get better results.

**Comment by smt9020** [comment]
I am having this same problem.  If you install version 1.0.0 it works.

**Comment by eskim0223** [comment]
Yes,installing version 1.0.0 also worked for me

**Comment by kibertoad** [comment]
I've just added a test that tries replicating this issue: https://github.com/foreversd/forever/pull/1068
It works fine both locally (Windows 10) and in CI.
If someone could provide a reliable repro for the problem, that would be fantastic.

That said, I'm pretty sure problem is caused by https://github.com/foreversd/forever-monitor/pull/145, but would be great if I could have a reproduction to confirm before doing any changes, that hiding that change behind an option would indeed help.

**Comment by davecarlile** [comment]
Windows 10:

> npm install forever-monitor

hello.js:
console.log('hello world')

index.js:
var forever = require('forever-monitor');
var child = new (forever.Monitor)('hello.js', {
    max: 3,
    silent: false,
    args: []
  });
 
  child.on('exit', function () {
    console.log('hello.js has exited after 3 restarts');
  });
 
  child.start();

> node index

result:
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
your-filename.js has exited after 3 restarts

Is there a workaround for this? Can I explicitly point to the node binary?

**Comment by sasodoma** [comment]
What worked for me, since node.exe is in my path, i just used forever start -c node app.js and it started correctly.

**Comment by foodbandlt** [comment]
@sasodoma I got to this same conclusion before finding this thread.  I ran into a problem doing it that way though, would you mind checking if you have the same problem?  

Once I started the daemon using -c node, when I used forever stopall, it failed to kill the child process for the script.  So forever list says nothing is running, but the process for the script is still running in the background.

I ended up downgrading as mentioned above to 1.0.1, but I'm curious if this was something others had issues with too.

**Comment by sasodoma** [comment]
@foodbandlt Yes I had the exact same problem. My app was running and express server and socket.io. Running stopall didn't kill it and I also couldn't restart it because it wouldn't kill the old one and the the ports would clash.

Also note that I don't think this is a problem with the -c option since I have tried the other way (using a short name in the PATH) and it still couldn't kill the process.

**Comment by yosiasz** [comment]
The way it worked for me, in Windows, was to choose a folder to install to that does not contain a space. So Installed node to c:\ProgramFiles instead of c:\Program Files ```=> {  Works! }```

**Comment by Boleroz** [comment]
Why not just enclose the path in quotes?
process.execPath = '"' + path.normalize(process.execPath) + '"';

**Comment by TuomoKu** [comment]
I'm having the same issue `C:\Program' is not recognized as an internal or external command..` when trying to use `forever` (v1.0.0) in npm start script.  The issue appears on my client's Win10 PC but not on my development Win10 PCs. Env path of Node is changed to short name `PROGR~1`which helps with _node_ but not with _forever_.

**Comment by kibertoad** [comment]
Can you try forever 3.0.0?

**Comment by Didier68** [comment]
Hi, I have the same issue with forever 2.0.0 on a Windows 10  PC... and I cannot upgrade "forever" because of an " No matching version found for forever-monitor@^4.0.1." error...
